File: src/noise/variablePersistenceMultioctaveNoise.ts
/**
 * A reimplementation of Factorio's `variable_persistence_multioctave_noise`
 * primitive (`NoiseOperations::VariablePersistenceMultioctaveNoise`),
 * reverse-engineered against Factorio 2.1.11 - by disassembling its register
 * `run` (`NoiseOperations::VariablePersistenceMultioctaveNoise::run` @0x10174a318)
 * and fitting the committed oracle. See
 * docs/noise/variable-persistence-multioctave-noise-NOTES.md. Built on
 * {@link basisNoise}.
 *
 * This is the op the **elevation** tree uses (nauvis `make_0_12like_lakes`). Its
 * defining feature: `persistence` is a spatially-varying value (a noise
 * *expression* the game evaluates per tile), so successive octaves are attenuated
 * by a persistence that changes across the map. Here the caller supplies that
 * per-tile `persistence` scalar.
 *
 * The shape:
 *
 *   varPers(x, y) = f32( gain * HORNER_{k=0..N-1} basis( f32(f32(x + offset_x)*S_k) ,
 *                                                        f32(y*S_k) ) )
 *
 *     S_0   = f32(input_scale * 0.5)     (finest octave scale = input_scale/2)
 *     S_k+1 = f32(S_k * 0.5)
 *     gain  = output_scale * 2^N
 *     p     = persistence at this tile
 *     N     = octaves
 *
 * i.e. N octaves of `basis_noise` sharing ONE (seed0, seed1) - each octave halves
 * the input scale (lacunarity 1/2) and is weighted by a power of the per-tile
 * persistence, combined in Horner order (finest octave gets the smallest weight
 * p^(N-1), coarsest gets 1), with every step rounded to f32. Two things
 * distinguish it from the relatives:
 *
 * 1. **No RMS normalisation.** It is the raw weighted sum times a `2^N` gain. The
 *    `amplitude_corrected_multioctave_noise` Lua wrapper is what normalises, by
 *    passing `output_scale = (1 - p)/2^N/(1 - p^N) * amplitude`. (The RMS-norm
 *    branch in the `Noise::multioctaveNoise(...,float const*,...)` float overload
 *    is a *different* entry point; the register `run` path the game evaluates via
 *    `calculate_tile_properties` has none.)
 * 2. **`offset_x` is a single world-space x translation** `(x + offset_x)*scale`,
 *    applied identically to every octave - like `quick_multioctave_noise`.
 *
 * **There is NO per-octave x shift.** This file used to carry a fitted
 * `k*(-7936)`; `::run`'s octave loop reloads the x/y offsets from the same two
 * constant slots every iteration and has no counter-scaled term at all. See the
 * comment on the removal below - the fitted value was an alias of zero.
 *
 * Verified against the game at **1.1e-5 worst over all 266 oracle samples**, and
 * that residual is `basisNoise`'s own f32 floor amplified by this op's gain, not a
 * modelling gap: `worst/gain` is 1.2e-7 to 2.4e-7 - one to two f32 ulps - in every
 * one of the seven cases, including the two with `offset_x` of 5000 and 40000.
 * NOT wired into the app - a building block for a client-side map preview.
 */

import { basisNoise, basisNoiseTablesFromSeed, type BasisNoiseTables } from "./basisNoise";

/**
 * `OCTAVE_SHIFT` is GONE, and its absence is the fix.
 *
 * It was `-7936`, fitted as "independent of seed, input_scale, offset_x and
 * persistence to the noise floor". The fit was not measuring a shift - it was
 * measuring nothing. The basis lattice has period 256 per axis, and
 * `-7936 == -31 * 256`, so it names the same field as a shift of **zero**. The
 * old comment's own "false minima near -4864 / -3840" are `-19*256` and `-15*256`
 * - every candidate the scan surfaced was a multiple of 256, which is what a
 * completely flat fit direction looks like from the inside.
 *
 * `VariablePersistenceMultioctaveNoise::run`'s octave loop settles it: it reloads
 * the x and y offsets from the same two constant slots (`+0xa2c`, `+0xa30`) on
 * every iteration and contains no counter-scaled term. There is no per-octave
 * shift; octaves decorrelate through lacunarity alone.
 *
 * In f64 removing it changes nothing (measured: identical worst and identical
 * f32-exact count). In f32 it is the difference between **3.6e-1 and 1.1e-5** -
 * a shift of -7936*5 lands where an f32 ulp is ~3.9e-3. Same defect and same
 * pairing as the plain op's `-1774.83`; see docs/noise/multioctave-noise-NOTES.md.
 */
const f32 = Math.fround;

export interface VariablePersistenceMultioctaveParams {
  /** Map seed (basis seed word). */
  readonly seed0: number;
  /** Per-call seed selector (distinguishes the many multioctave calls a program makes). */
  readonly seed1: number;
  /** Octave count (>= 1). */
  readonly octaves: number;
  /** Base input scale (noise units per world tile); octave 0 uses `input_scale/2`. */
  readonly inputScale: number;
  /** Overall output multiplier (the op additionally applies a `2^octaves` gain). */
  readonly outputScale: number;
  /** World-space x translation applied to every octave (`(x + offsetX)` before scaling). */
  readonly offsetX: number;
}

/**
 * Evaluate `variable_persistence_multioctave_noise` at world coordinates `(x, y)`
 * with a per-tile `persistence`. Pass a prebuilt `tables` to skip the seed
 * derivation when sweeping many points at one seed (the common case for rendering);
 * the seed is shared across octaves, so a single tables object serves them all.
 *
 * The octaves are combined in Horner order exactly as the game's `run` does (add
 * the octave, then multiply the running accumulator by the tile's persistence -
 * except after the last octave), so octave k carries weight `p^(N-1-k)`.
 */
export function variablePersistenceMultioctaveNoise(
  x: number,
  y: number,
  persistence: number,
  params: VariablePersistenceMultioctaveParams,
  tables: BasisNoiseTables = basisNoiseTablesFromSeed(params.seed0, params.seed1),
): number {
  const { octaves, inputScale, outputScale, offsetX } = params;

  let acc = 0;
  let scale = f32(f32(inputScale) * 0.5); // octave 0 = input_scale / 2
  for (let k = 0; k < octaves; k++) {
    acc = f32(acc + basisNoise(f32(f32(x + offsetX) * scale), f32(y * scale), tables));
    if (k < octaves - 1) acc = f32(acc * persistence);
    scale = f32(scale * 0.5);
  }
  return f32(acc * f32(outputScale * 2 ** octaves));
}

/**
 * Build a closure that evaluates `variable_persistence_multioctave_noise` for a
 * fixed parameter set, with the per-octave scales (and the shared basis tables)
 * derived once up front (the common case for rendering a grid at one seed). The
 * returned function takes `(x, y, persistence)` - persistence still varies per tile.
 */
export function makeVariablePersistenceMultioctaveNoise(
  params: VariablePersistenceMultioctaveParams,
): (x: number, y: number, persistence: number) => number {
  const { seed0, seed1, octaves, inputScale, outputScale, offsetX } = params;
  const tables = basisNoiseTablesFromSeed(seed0, seed1);
  const gain = f32(outputScale * 2 ** octaves);

  const octaveScale: number[] = [];
  let scale = f32(f32(inputScale) * 0.5);
  for (let k = 0; k < octaves; k++) {
    octaveScale.push(scale);
    scale = f32(scale * 0.5);
  }

  return (x: number, y: number, persistence: number): number => {
    let acc = 0;
    for (let k = 0; k < octaves; k++) {
      const s = octaveScale[k];
      acc = f32(acc + basisNoise(f32(f32(x + offsetX) * s), f32(y * s), tables));
      if (k < octaves - 1) acc = f32(acc * persistence);
    }
    return f32(acc * gain);
  };
}

export interface AmplitudeCorrectedMultioctaveParams {
  /** Map seed (basis seed word). */
  readonly seed0: number;
  /** Per-call seed selector. */
  readonly seed1: number;
  /** Octave count (>= 1). */
  readonly octaves: number;
  /** Base input scale (noise units per world tile); octave 0 uses `input_scale/2`. */
  readonly inputScale: number;
  /** World-space x translation applied to every octave. */
  readonly offsetX: number;
  /** Constant persistence (amplitude ratio between successive octaves). */
  readonly persistence: number;
  /** Target output amplitude (the corrected sum is scaled to roughly this). */
  readonly amplitude: number;
}

/**
 * `amplitude_corrected_multioctave_noise` - the Lua wrapper
 * (`core/prototypes/noise-functions.lua`) over
 * {@link variablePersistenceMultioctaveNoise}. It just chooses the op's
 * `output_scale` so the `2^N`-gained geometric sum ends up at roughly `amplitude`:
 *
 *   output_scale = (1 - p) / 2^N / (1 - p^N) * amplitude
 *
 * (the `1 - p^N` is the geometric-series sum of the octave weights; the `/2^N`
 * cancels the op's gain). The wrapper's `persistence` is a constant here - and the
 * game's degenerate constant-persistence path yields the same math as the variable
 * op (verified against the oracle to the basis floor), so this is a direct call.
 *
 * At `p == 1` the `(1 - p)/(1 - p^N)` ratio is 0/0; its limit is `1/N`, so
 * `output_scale = amplitude / (N * 2^N)`.
 *
 * The elevation tree uses this to build the *persistence field* it then feeds to
 * `make_0_12like_lakes` (`persistence = clamp(amplitude_corrected... + 0.3, 0.1, 0.9)`).
 */
export function amplitudeCorrectedMultioctaveNoise(
  x: number,
  y: number,
  params: AmplitudeCorrectedMultioctaveParams,
  tables: BasisNoiseTables = basisNoiseTablesFromSeed(params.seed0, params.seed1),
): number {
  const { octaves, persistence: p, amplitude } = params;
  const ratio = p === 1 ? 1 / octaves : (1 - p) / (1 - p ** octaves);
  const outputScale = (ratio / 2 ** octaves) * amplitude;
  return variablePersistenceMultioctaveNoise(
    x,
    y,
    p,
    {
      seed0: params.seed0,
      seed1: params.seed1,
      octaves,
      inputScale: params.inputScale,
      outputScale,
      offsetX: params.offsetX,
    },
    tables,
  );
}

File: crates/fmw-noise/src/variable_persistence_multioctave_noise.rs
//! `variable_persistence_multioctave_noise` -
//! `NoiseOperations::VariablePersistenceMultioctaveNoise`.
//!
//! Ported from `src/noise/variablePersistenceMultioctaveNoise.ts`; derivation
//! in `docs/noise/variable-persistence-multioctave-noise-NOTES.md`. This is the
//! op the elevation tree uses (nauvis `make_0_12like_lakes`). Its defining
//! feature: `persistence` is spatially varying - a noise expression the game
//! evaluates per tile - so the caller supplies it per point.
//!
//! N octaves of `basis_noise` sharing one `(seed0, seed1)`, each halving the
//! input scale and weighted by a power of the per-tile persistence, combined in
//! Horner order so octave k carries weight `p^(N-1-k)`, every step rounded to
//! f32. No RMS normalisation - it is the raw weighted sum times a `2^N` gain.
//!
//! **There is NO per-octave x shift.** The port once carried a fitted `-7936`,
//! which is `-31 * 256` and so names the same field as zero; `::run`'s octave
//! loop reloads the x/y offsets from the same two constant slots every
//! iteration and has no counter-scaled term. In f64 removing it changes
//! nothing; in f32 it is the difference between 3.6e-1 and bit-exact.

use crate::basis_noise::{basis_noise, tables_from_seed, BasisNoiseTables};

/// Parameters for [`variable_persistence_multioctave_noise`]. `f64` for the
/// same reason as [`crate::multioctave_noise::MultioctaveParams`].
#[derive(Clone, Copy, Debug)]
pub struct VariablePersistenceParams {
    /// Map seed.
    pub seed0: u32,
    /// Per-call seed selector.
    pub seed1: u32,
    /// Octave count (>= 1).
    pub octaves: u32,
    /// Base input scale; octave 0 uses `input_scale / 2`.
    pub input_scale: f64,
    /// Overall output multiplier (the op additionally applies a `2^octaves` gain).
    pub output_scale: f64,
    /// World-space x translation applied to every octave.
    pub offset_x: f64,
}

/// The per-octave scales and the folded output gain.
pub struct VariablePersistenceTerms {
    scales: Vec<f32>,
    gain: f32,
    offset_x: f64,
}

/// Derive the per-octave input scales and the `output_scale * 2^N` gain.
#[must_use]
pub fn terms(params: &VariablePersistenceParams) -> VariablePersistenceTerms {
    // `2^octaves` is exact in f64 for every reachable octave count, so the
    // product is one f64 multiply narrowed once - matching the TypeScript,
    // where both operands are JavaScript numbers.
    let gain = (params.output_scale * f64::from(2u32.pow(params.octaves))) as f32;

    let mut scale = (params.input_scale as f32) * 0.5;
    let mut scales = Vec::with_capacity(params.octaves as usize);
    for _ in 0..params.octaves {
        scales.push(scale);
        scale *= 0.5;
    }
    VariablePersistenceTerms {
        scales,
        gain,
        offset_x: params.offset_x,
    }
}

/// Evaluate the op at `(x, y)` with a per-tile `persistence`.
///
/// Horner order exactly as the game's `run` does it: add the octave, then
/// multiply the running accumulator by the tile's persistence - except after
/// the last octave.
///
/// Note the two coordinates are NOT symmetric, and that mirrors the
/// TypeScript rather than being an oversight. `x + offset_x` is a sum of two
/// f64s narrowed to f32 before the scale multiply, while `y * scale` is an f64
/// multiply of the un-narrowed `y` that is narrowed afterwards. Making them
/// match would change which point is evaluated.
///
/// ## `persistence` is f64, and it took a NEW fixture to find that out (#226)
///
/// It was `f32` from phase 1 until phase 6, matching neither the TypeScript -
/// where `acc = f32(acc * persistence)` multiplies an f32 accumulator by an
/// un-narrowed JavaScript number - nor, as it turns out, the game.
///
/// **No fixture in the tree could see it, and that is measured rather than
/// argued.** `oracle-variable-persistence-multioctave`'s captured
/// `persistenceField` is `0.35 + 0.25 * basis_noise{...}` evaluated in the
/// noise machine, so all 38 of its values are exactly f32; at an f32-exact
/// persistence the two forms are bit-identical, and that test scores 152/152
/// either way. It is the same shape as #191 and #309: a narrowing the fixtures
/// agree on because the fixtures only ever offer values already on the grid.
///
/// What DOES discriminate is `oracle-multioctave-wrappers`'s
/// `amplitudeCorrected` section, because those cases pass a raw program
/// constant - `0.7`, which is not f32-exact - straight in as the persistence.
/// Scored against the game over its 152 points:
///
/// ```text
/// persistence held at f64 (the TypeScript, and now this)   81/152, worst 1.788e-7
/// persistence narrowed to f32 (what shipped here)          89/152, worst 5.960e-8
/// ```
///
/// **The better-scoring form is NOT the one that ships**, and that is
/// deliberate. 89 is an improvement and not a full exact count, so the
/// greedy-accept rule rejects it as a model change; and the port's standing
/// rule is to reproduce the TypeScript so tier 2 compares two implementations
/// of one model rather than two models. Taking the 8 points here would put a
/// divergence into every Nauvis elevation value with nothing to grade it.
///
/// It is a real finding for #254 all the same: that issue records the
/// amplitude-corrected wrapper's 81/152 as unexplained, and this says the
/// persistence operand's width is one term in it, worth 8 points and still 63
/// short. Neither form is the game's.
#[must_use]
pub fn eval(
    x: f64,
    y: f64,
    persistence: f64,
    terms: &VariablePersistenceTerms,
    tables: &BasisNoiseTables,
) -> f32 {
    let xo = (x + terms.offset_x) as f32;
    // `y` is narrowed for the same reason `x` is: the noise machine holds its
    // coordinate values at f32, so the scale multiply is an f32 operation on
    // both operands. `x` was always narrowed here through the `offset_x` add;
    // `y` had no add to narrow it and so was silently multiplied in f64 (#191).
    let yf = y as f32;
    let last = terms.scales.len().saturating_sub(1);
    let mut acc = 0.0f32;
    for (k, &scale) in terms.scales.iter().enumerate() {
        let xk = xo * scale;
        let yk = yf * scale;
        acc += basis_noise(f64::from(xk), f64::from(yk), tables);
        if k < last {
            // f32 accumulator times an f64 persistence, narrowed once - the
            // TypeScript's `f32(acc * persistence)`. Widening `acc` here rather
            // than narrowing `persistence` is the whole content of the note
            // above; do not "simplify" it back to an f32 multiply.
            acc = (f64::from(acc) * persistence) as f32;
        }
    }
    acc * terms.gain
}

/// Evaluate `variable_persistence_multioctave_noise`, deriving tables and terms
/// per call. Use [`terms`] plus [`eval`] when sweeping a grid.
#[must_use]
pub fn variable_persistence_multioctave_noise(
    x: f64,
    y: f64,
    persistence: f64,
    params: &VariablePersistenceParams,
) -> f32 {
    let tables = tables_from_seed(params.seed0, params.seed1);
    eval(x, y, persistence, &terms(params), &tables)
}

/// Free variables of `amplitude_corrected_multioctave_noise`.
pub struct AmplitudeCorrectedParams {
    /// Map seed (basis seed word).
    pub seed0: u32,
    /// Per-call seed selector.
    pub seed1: u32,
    /// Octave count.
    pub octaves: u32,
    /// Noise units per world tile.
    pub input_scale: f64,
    /// World-space x translation applied before `input_scale`.
    pub offset_x: f64,
    /// The persistence the correction is computed for, and the one the octave
    /// stack below runs at.
    pub persistence: f64,
    /// Target amplitude the correction normalises to.
    pub amplitude: f64,
}

/// `amplitude_corrected_multioctave_noise`, ported from
/// `amplitudeCorrectedMultioctaveNoise` in
/// `src/noise/variablePersistenceMultioctaveNoise.ts`.
///
/// A thin wrapper: it derives an `output_scale` that normalises a
/// variable-persistence stack to a target amplitude, then delegates. The
/// geometric series `(1 - p) / (1 - p^octaves)` is the sum of the octave
/// amplitudes, with the `p == 1` branch standing in for the removable
/// singularity there.
///
/// ## It is 81 of 152 against the game, and that is #254
///
/// The op underneath is bit-exact on the same fixture (152/152), so the gap is
/// in this transform. It is ported EXACTLY as the TypeScript writes it, in f64,
/// rather than corrected - the port's standing rule, so tier 2 compares two
/// implementations of one model instead of two different models.
///
/// Its sibling `quick_multioctave_noise_persistence` had the same shape of gap
/// and reaching 152/152 needed the wrapper's transform run in the noise
/// machine's f32. **That treatment does NOT fix this one** - measured on the
/// TypeScript side at 84/152, no better than the f64 form - so the obvious
/// candidate is already refuted and this is not a matter of applying it here.
///
/// ## `p^octaves` is `powi`, and the bits were checked against V8
///
/// The exponent is a small whole number and the base is an f64, so this is a
/// place where two libms could disagree by an ULP the way #270's `log2`/`pow`
/// did. `powi` is not obviously the same function as V8's `**`.
///
/// Rather than assume, `the_geometric_series_matches_v8s_pow_bit_for_bit` pins
/// the raw bits of `p^n` at every pair this port can reach against the values
/// V8 produces. If a future toolchain changes `powi`, that test names it
/// instead of a Nauvis elevation count moving for no visible reason.
#[must_use]
pub fn amplitude_corrected_multioctave_noise(
    x: f64,
    y: f64,
    params: &AmplitudeCorrectedParams,
) -> f32 {
    let p = params.persistence;
    let octaves = params.octaves;
    let ratio = if p == 1.0 {
        1.0 / f64::from(octaves)
    } else {
        (1.0 - p) / (1.0 - p.powf(f64::from(octaves)))
    };
    let output_scale = (ratio / f64::from(2u32.pow(octaves))) * params.amplitude;
    variable_persistence_multioctave_noise(
        x,
        y,
        p,
        &VariablePersistenceParams {
            seed0: params.seed0,
            seed1: params.seed1,
            octaves,
            input_scale: params.input_scale,
            output_scale,
            offset_x: params.offset_x,
        },
    )
}

#[cfg(test)]
mod amplitude_corrected_tests {
    #[test]
    fn the_geometric_series_matches_v8s_pow_bit_for_bit() {
        // Captured from V8 with `new DataView(...).setFloat64(0, p ** n)`, so
        // these are the exact f64s the TypeScript's transform divides by. Every
        // (persistence, octaves) pair this port can reach is here: the four
        // `oracle-multioctave-wrappers` cases, plus `elevation_lakes` (0.7 at 6)
        // and `elevation_nauvis` (0.7 at 5), plus the `p == 1` branch's guard.
        for (p, n, bits) in [
            (0.7f64, 2u32, 0x3fdf_5c28_f5c2_8f5bu64),
            (0.7, 3u32, 0x3fd5_f3b6_45a1_cabf),
            (0.7, 4, 0x3fce_bb98_c7e2_823f),
            (0.7, 5, 0x3fc5_8351_58b8_27f8),
            (0.7, 6, 0x3fbe_1e3e_af68_37f5),
            (0.7, 8, 0x3fad_840a_3b42_4b50),
            (0.5, 4, 0x3fb0_0000_0000_0000),
            (0.6, 2, 0x3fd7_0a3d_70a3_d70a),
            (1.0, 5, 0x3ff0_0000_0000_0000),
        ] {
            assert_eq!(
                p.powf(f64::from(n)).to_bits(),
                bits,
                "powf({p}, {n}) differs from V8's `**` - see #270 for what an \
                 ULP of disagreement here costs"
            );
        }
    }

    #[test]
    fn the_persistence_one_branch_is_the_series_limit_and_not_a_guess() {
        // At `p == 1` the closed form is 0/0. The branch returns `1 / octaves`,
        // which is what the series converges to - checked by approaching it,
        // so a branch returning some other constant could not pass.
        for octaves in [2u32, 5, 8] {
            let n = f64::from(octaves);
            let near = |p: f64| (1.0 - p) / (1.0 - p.powf(f64::from(octaves)));
            let limit = 1.0 / n;
            assert!(
                (near(1.0 - 1e-9) - limit).abs() < 1e-6,
                "the series does not approach 1/{n} at {octaves} octaves"
            );
        }
    }
}
